### PR TITLE
Push buffer

### PR DESCRIPTION
**Pull request by PatrickvL**
This pull request attempts to render push-buffers that have a so-called "non-FVF" vertex shader active (for now, only supported shaders are the ones that use just stream zero).

### CODE REVIEW

**Comment by Voxel9** [comment]
Leisure Suit Larry MCL no longer gets in-game with this PR, ends up crashing when the loading screen gets to 99%:

```
---------------------------
Cxbx-Reloaded
---------------------------
Received Exception Code 0xC0000005 @ EIP := 0x04087490(=XID_fCloseDevice+0x3e460ec)

  Press "OK" to terminate emulation.
  Press "Cancel" to debug.
---------------------------
OK   Cancel   
---------------------------
```

[KrnlDebug.txt](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/2066221/KrnlDebug.txt)

**Comment by PatrickvL** [comment]
@Voxel9 I've updated push buffer command decoding and handling - could you give it a quick test again?

**Comment by Voxel9** [comment]
Unfortunately I'm still getting the same error 😕

```---------------------------
Cxbx-Reloaded
---------------------------
Received Exception Code 0xC0000005 @ EIP := 0x04094470(=XID_fCloseDevice+0x3e530cc)

  Press "OK" to terminate emulation.
  Press "Cancel" to debug.
---------------------------
OK   Cancel   
---------------------------
```

[KrnlDebug.txt](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/2068709/KrnlDebug.txt)

**Comment by PatrickvL** [comment]
Behaviour did change though. The log shows a return command was hit (while previously a call failed)

**Comment by GXTX** [comment]
Lots of test cases on Hot Wheels Stunt Track Challenge while running `hw2F.xbe`. Doesn't show the logo xmv's but does show the main menu (minus text). Crashes while on the menu after hitting start a few times.

```
LOG_TEST_CASE: Pushbuffer jump and call?!
LOG_TEST_CASE: Pushbuffer instruction unknown!
LOG_TEST_CASE: Pushbuffer jump
LOG_TEST_CASE: Pushbuffer subchannel > 0
LOG_TEST_CASE: Pushbuffer call return with additional bits?!
```

**Comment by PatrickvL** [comment]
@Voxel9 @GXTX I've done some more work on this - would you care to re-test??

PS: Thanks for the mention of Metal Arms, finally a titles I've got!

**Comment by GXTX** [comment]
Metal Arms is now very very slow whereas before it was ~60FPS, Hot Wheels still crashes with test cases of 
```
LOG_TEST_CASE: Pushbuffer COMMAND_TYPE_JUMP
LOG_TEST_CASE: Pushbuffer COMMAND_INSTRUCTION unknown
LOG_TEST_CASE: Pushbuffer COMMAND_FLAGS_LONG_NON_INCREASING_METHODS [IB-mode only] not available on NV2A
LOG_TEST_CASE: Pushbuffer COMMAND_INSTRUCTION_OLD_JUMP
LOG_TEST_CASE: Pushbuffer COMMAND_TYPE unknown
LOG_TEST_CASE: Pushbuffer COMMAND_FLAGS_SLI_CONDITIONAL (NV40+) not available on NV2A
LOG_TEST_CASE: Pushbuffer COMMAND_FLAGS_RETURN with additional bits?!
```

**Comment by PatrickvL** [comment]
What a mess!! But thanks for testing

**Comment by ObiKKa** [comment]
That's a lot of new test case messages. How will you post the issues for them all? Group them up into a general Pushbuffer test cases issue?

File: src/CxbxKrnl/CxbxKrnl.h [hunk 1]
```diff
@@ -237,10 +237,18 @@ typedef enum _CxbxMsgDlgIcon {
 
 void CxbxPopupMessage(CxbxMsgDlgIcon icon, const char *message, ...);
 
+#ifdef _DEBUG
+#define LOG_TEST_CASE(message) do { static bool bTestCaseLogged = false; \
+    if (!bTestCaseLogged) { bTestCaseLogged = true; \
```
Comment (LukeUsher) [inline src/CxbxKrnl/CxbxKrnl.h]: Could you also output the printf format as well as showing the dialog?
The printf format with the LOG_TEST_CASE prefix will be automatically crawled and indexed by the website.

Comment (PatrickvL) [reply src/CxbxKrnl/CxbxKrnl.h]: I believe the CxbxPopup function already does that

Comment (PatrickvL) [reply src/CxbxKrnl/CxbxKrnl.h]: I've checked - 	`CxbxPopupMessage` calls `EmuWarning("Popup : %s\n", Buffer);` so anything shown in a popup will also show up in the log.
With that, all that's left to do is to make sure the `LOG_TEST_CASE` sub-string is in there - I'll adjust the macro for that.

**Comment by PatrickvL** [comment]
@Voxel9 @GXTX Okay guys - another round of refactoring, another round of testing...

I've tested DEBUG and RELEASE builds with BeginPush, Freedom Fighters, Metal Arms and Turok - no regressions that I could see (compare to before).

**Comment by Voxel9** [comment]
Leisure Suit Larry gets back in-game with the largely-flickering graphics, though compared to master it doesn't clear the screen properly.

[KrnlDebug.txt](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/2081449/KrnlDebug.txt)

image

**Comment by LukeUsher** [comment]
Excellent work!

File: src/Cxbx.h [hunk 2]
```diff
@@ -79,8 +79,6 @@ typedef u32              xbaddr;
 #define _DEBUG_TRACK_VS
 /*! define this to track pixel shaders */
 #define _DEBUG_TRACK_PS
-/*! define this to track push buffers */
-#define _DEBUG_TRACK_PB
```
Comment (RadWolfie) [inline src/Cxbx.h]: Is removing this define intentional?

Comment (PatrickvL) [reply src/Cxbx.h]: Yes. I've removed all push buffer debugging code that didn't work anymore anyway

**Comment by PatrickvL** [comment]
@Voxel9 I suspect the flickering graphics are caused by a non-FVF vertex shader being active during push-buffer rendering. Could you please create a separate issue for that, including the above screenshot as an example? (Hopefully, we can find more test-cases)

**Pull request merged**